app/models/order.rb

```diff
@@ -1,12 +1,65 @@
 class Order < ApplicationRecord
-  self.state_machine({
-    cart: [:placed, :canceled],
-    placed: [:canceled],
-    canceled: []
-  })
+  attr_accessor :credit_card_id
 
   validates :state, presence: true
   belongs_to :user
   has_many :order_items, as: :source
-  #todo implement state transition logic
+  has_one :payment
+
+  self.state_machine({
+       cart: [:placed],
+       placed: [:cancelled],
+       cancelled: []
+   })
+
+  before_transition_to :placed do
+    if order_items.to_a.sum {|x| x.quantity} == 0
+      errors[:base] << "Order must contain at least one item"
+    end
+
+    order_items.each do |item|
+      if item.source.on_hand_count < item.quantity
+        errors[:base] << "There is not enough #{item.name} in inventory to complete your order."
+      end
+    end
+
+    #consume the orders nom nom nom <(-.-<)
+    order_items.each do |item|
+      item.source.on_hand_count -= item.quantity
+      item.source.available = item.on_hand_count > 0
+      unless item.source.save
+        errors[:base] << "Unable to properly save #{item.source.name}"
+      end
+    end
+
+    payment = Payment.create(credit_card_id: credit_card_id, order_id: self.id, state: pending)
+    payment.amount = order_total
+
+    unless payment.make_placed
+      error[:payment] << "Error changing payment state to completed"
+    end
+    unless payment.save
+      error[:payment] << "Unable to properly save payment"
+    end
+    self.total = order_items.to_a.map(&:price).inject(0, &:+)
```

Comment:
I like the use of inject! but it could just be
```ruby
self.total = order_items.to_a.sum(&:price)
```
unless the default for price on order_item is nil instead of 0.